util/auto_tune_compressor.h

```diff
@@ -140,50 +143,87 @@ class CostAwareWorkingArea : public Compressor::WorkingArea {
-  Status CompressBlockAndRecord(size_t choosen_compression_type,
-                                size_t compresion_level_ptr,
+  Status CompressBlockAndRecord(size_t compressor_index,
                                 Slice uncompressed_data,
                                 std::string* compressed_output,
                                 CompressionType* out_compression_type,
                                 CostAwareWorkingArea* wa);
+
+  inline bool IsInValidQuadrant(size_t predicted_io_cost,
+                                size_t predicted_cpu_cost, size_t cur_io_cost,
+                                size_t cur_cpu_cost, bool increase_io,
+                                bool increase_cpu, bool decrease_io,
+                                bool decrease_cpu) {
+    return (!increase_io || predicted_io_cost > cur_io_cost) &&
+           (!decrease_io || predicted_io_cost < cur_io_cost) &&
+           (!increase_cpu || predicted_cpu_cost > cur_cpu_cost) &&
+           (!decrease_cpu || predicted_cpu_cost < cur_cpu_cost);
+  }
+  // Select compression type and level based on budget availability
+  size_t SelectCompressionBasedOnIOGoalCPUBudget(CostAwareWorkingArea* wa);
+  void MeasureUtilization();
+  void AddCompressors(CompressionType type,
+                      const std::initializer_list<int>& levels);
+  static constexpr uint64_t kMicrosInSecond = 1000000;
   static constexpr int kExplorationPercentage = 10;
   static constexpr int kProbabilityCutOff = 50;
-  // This is the vector containing the list of compression levels that
-  // CostAwareCompressor will use create compressor and predicts the cost
-  // The vector contains list of compression level for compression algorithm in
-  // the order defined by enum CompressionType
-  static const std::vector<std::vector<int>> kCompressionLevels;
   const CompressionOptions opts_;
-  std::vector<std::vector<std::unique_ptr<Compressor>>> allcompressors_;
-  std::vector<std::pair<size_t, size_t>> allcompressors_index_;
+
+  // Budget references for cost-aware compression decisions
+  std::shared_ptr<IOGoal> io_goal_;
+  std::shared_ptr<CPUBudget> cpu_budget_;
+  CPUIOUtilizationTracker usage_tracker_;
+  // Will serve as a logical clock to decide when to update the decision
+  std::atomic<int> block_count_;
+  std::atomic<size_t> cur_compressor_idx_;
+  std::unique_ptr<Compressor> default_compressor_;
+
+  static constexpr int kCompressionEvaluationInterval = 2000;
+  static constexpr int kWindow = 10;
 };
 
-class CostAwareCompressorManager : public CompressionManagerWrapper {
-  using CompressionManagerWrapper::CompressionManagerWrapper;
+class AutoTuneCompressorManager : public CompressionManagerWrapper {
+ public:
+  explicit AutoTuneCompressorManager(
+      std::shared_ptr<CompressionManager> wrapped,
+      std::shared_ptr<IOGoal> io_goal, std::shared_ptr<CPUBudget> cpu_budget,
+      const Options& option)
+      : CompressionManagerWrapper(wrapped),
+        io_goal_(io_goal),
```

Comment:
```suggestion
        io_goal_(std::move(io_goal),
```
or pass in io_goal by reference.